internal/cleanup/cleanup.go

```diff
@@ -57,54 +58,52 @@ type exposureCleanupHandler struct {
-	ctx, span := trace.StartSpan(r.Context(), "(*cleanup.exposureCleanupHandler).ServeHTTP")
-	defer span.End()
-
-	logger := logging.FromContext(ctx)
+	ctx := r.Context()
+	logger := logging.FromContext(ctx).Named("cleanup.exposure")
 
-	cutoff, err := cutoffDate(ctx, h.config.TTL, h.config.DebugOverrideCleanupMinDuration)
+	cutoff, err := cutoffDate(h.config.TTL, h.config.DebugOverrideCleanupMinDuration)
 	if err != nil {
-		message := fmt.Sprintf("error processing cutoff time: %v", err)
-		logger.Error(message)
-		span.SetStatus(trace.Status{Code: trace.StatusCodeInternal, Message: message})
-		stats.Record(ctx, mExposuresSetupFailed.M(1))
-		http.Error(w, "internal processing error", http.StatusInternalServerError)
+		logger.Errorw("failed to calculate cutoff date", "error", err)
+		respondWithError(w, http.StatusInternalServerError, err)
 		return
 	}
-	logger.Infof("Starting cleanup for records older than %v", cutoff.UTC())
-	stats.Record(ctx, mExposuresCleanupBefore.M(cutoff.Unix()))
 
-	// Set timeout
-	timeoutCtx, cancel := context.WithTimeout(ctx, h.config.Timeout)
-	defer cancel()
+	// Construct a multi-error. If one of the purges fails, we still want to
+	// attempt the other purges.
+	var merr *multierror.Error
 
-	count, err := h.database.DeleteExposuresBefore(timeoutCtx, cutoff)
-	if err != nil {
-		message := fmt.Sprintf("Failed deleting exposures: %v", err)
-		logger.Error(message)
-		stats.Record(ctx, mExposuresDeleteFailed.M(1))
-		span.SetStatus(trace.Status{Code: trace.StatusCodeInternal, Message: message})
-		http.Error(w, "internal processing error", http.StatusInternalServerError)
-		return
-	}
+	// Exposures
+	func() {
+		ctx, cancel := context.WithTimeout(ctx, h.config.Timeout)
+		defer cancel()
 
-	stats.Record(ctx, mExposuresDeleted.M(count))
-	logger.Infof("cleanup run complete, deleted %v records.", count)
+		if count, err := h.database.DeleteExposuresBefore(ctx, cutoff); err != nil {
+			merr = multierror.Append(merr, fmt.Errorf("failed to delete exposures: %w", err))
+		} else {
+			logger.Infow("purged exposures", "count", count)
+		}
+	}()
 
-	// Clean up associated stats.
-	statsCount, err := h.database.DeleteStatsBefore(timeoutCtx, cutoff)
-	if err != nil {
-		message := fmt.Sprintf("Failed deleting publish stats: %v", err)
-		logger.Error(message)
-		stats.Record(ctx, mExposuresStatsDeleteFailed.M(1))
-		span.SetStatus(trace.Status{Code: trace.StatusCodeInternal, Message: message})
-		http.Error(w, "internal processing error", http.StatusInternalServerError)
+	// Stats
+	func() {
+		ctx, cancel := context.WithTimeout(ctx, h.config.Timeout)
+		defer cancel()
+
+		if count, err := h.database.DeleteStatsBefore(ctx, cutoff); err != nil {
+			merr = multierror.Append(merr, fmt.Errorf("failed to delete stats: %w", err))
+		} else {
+			logger.Infow("purged statistics", "count", count)
+		}
+	}()
+
+	if merr != nil {
+		logger.Errorw("failed to cleanup exposures", "errors", merr.WrappedErrors())
```

Comment:
log message needs updating, it's not just exposures now

Reply:
The underlying `errors` are still wrapped with the context of whether it was exposures or stats that failed